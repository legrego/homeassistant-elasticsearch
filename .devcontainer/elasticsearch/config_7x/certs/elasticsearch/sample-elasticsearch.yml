#
# SAMPLE ELASTICSEARCH CONFIGURATION FOR ENABLING SSL ON THE HTTP INTERFACE
#
# This is a sample configuration snippet for Elasticsearch that enables and configures SSL for the HTTP (Rest) interface
#
# This was automatically generated at: 2020-12-21 13:19:13-05:00
# This configuration was intended for Elasticsearch version 8.0.0
#
# You should review these settings, and then update the main configuration file at
#   /Users/larry/repos/kibana/.es/8.0.0/config/elasticsearch.yml
#

# This turns on SSL for the HTTP (Rest) interface
xpack.security.http.ssl.enabled: true

# This configures the keystore to use for SSL on HTTP
xpack.security.http.ssl.keystore.path: "http.p12"

