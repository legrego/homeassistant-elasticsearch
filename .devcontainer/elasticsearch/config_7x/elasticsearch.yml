#
# Path to directory where to store the data (separate multiple locations by comma):
#
path.data: /var/lib/elasticsearch
#
# Path to log files:
#
path.logs: /var/log/elasticsearch


cluster.name: home-assistant-cluster

xpack.security.http.ssl.enabled: true
xpack.security.http.ssl.keystore.path: certs/elasticsearch/http.p12
xpack.security.enabled: true