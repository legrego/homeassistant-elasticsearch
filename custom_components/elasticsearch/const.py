""" constants """

DOMAIN = "elasticsearch"

CONF_PUBLISH_ENABLED = "publish_enabled"
CONF_HEALTH_SENSOR_ENABLED = "health_sensor_enabled"
CONF_INDEX_FORMAT = "index_format"
CONF_PUBLISH_FREQUENCY = "publish_frequency"
CONF_EXCLUDED_DOMAINS = "excluded_domains"
CONF_EXCLUDED_ENTITIES = "excluded_entities"
CONF_PUBLISH_MODE = "publish_mode"
CONF_INCLUDED_DOMAINS = "included_domains"
CONF_INCLUDED_ENTITIES = "included_entities"
CONF_ONLY_PUBLISH_CHANGED = "only_publish_changed"
CONF_ILM_ENABLED = "ilm_enabled"
CONF_ILM_POLICY_NAME = "ilm_policy_name"
CONF_ILM_MAX_SIZE = "ilm_max_size"
CONF_ILM_DELETE_AFTER = "ilm_delete_after"
CONF_SSL_CA_PATH = "ssl_ca_path"

CONF_TAGS = "tags"

ONE_MINUTE = 60
ONE_HOUR = 60 * 60

VERSION_SUFFIX = "-v4_1"

INDEX_TEMPLATE_NAME = "hass-index-template" + VERSION_SUFFIX

PUBLISH_MODE_ALL = "All"
PUBLISH_MODE_STATE_CHANGES = "State changes"
PUBLISH_MODE_ANY_CHANGES = "Any changes"
