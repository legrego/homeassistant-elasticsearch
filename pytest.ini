[pytest]
addopts = -rxf -x -v -l --cov=./ --cov-report=xml
filterwarnings =
    ignore::DeprecationWarning
testpaths = tests
python_files =
    test_*.py