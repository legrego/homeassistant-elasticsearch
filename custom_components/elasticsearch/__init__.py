"""Support for sending event data to an Elasticsearch cluster."""

from __future__ import annotations

from logging import Logger
from typing import TYPE_CHECKING

from homeassistant.config_entries import ConfigEntry
from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady, IntegrationError

from custom_components.elasticsearch.config_flow import ElasticFlowHandler
from custom_components.elasticsearch.errors import (
    AuthenticationRequired,
    CannotConnect,
    ESIntegrationException,
    UnsupportedVersion,
)
from custom_components.elasticsearch.logger import (
    LOGGER,
    async_log_enter_exit_debug,
    async_log_enter_exit_info,
    have_child,
    log_enter_exit_debug,
)

from .es_integration import ElasticIntegration

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant

type ElasticIntegrationConfigEntry = ConfigEntry[ElasticIntegration]


@async_log_enter_exit_info
async def async_setup_entry(hass: HomeAssistant, config_entry: ElasticIntegrationConfigEntry) -> bool:
    """Set up integration via config flow."""

    # Create an specific logger for this config entry
    _logger: Logger = have_child(name=config_entry.title)

    _logger.info("Initializing integration for %s", config_entry.title)

    try:
        integration = ElasticIntegration(hass=hass, config_entry=config_entry, log=_logger)
        await integration.async_init()
    except (UnsupportedVersion, CannotConnect) as err:
        raise ConfigEntryNotReady(err) from err
    except AuthenticationRequired as err:
        raise ConfigEntryAuthFailed(err) from err
    except ESIntegrationException as err:
        raise ConfigEntryNotReady(err) from err
    except Exception as err:
        msg = "Unknown error occurred"
        _logger.exception(msg)
        raise IntegrationError(err) from err

    config_entry.runtime_data = integration
    return True


@async_log_enter_exit_info
async def async_unload_entry(hass: HomeAssistant, config_entry: ElasticIntegrationConfigEntry) -> bool:
    """Teardown integration."""

    if (
        hasattr(config_entry, "runtime_data")
        and config_entry.runtime_data is not None
        and isinstance(config_entry.runtime_data, ElasticIntegration)
    ):
        integration = config_entry.runtime_data

        await integration.async_shutdown()
    else:
        LOGGER.warning(
            "Called to unload config entry %s, but it doesn't appear to be loaded", config_entry.title
        )

    return True


@async_log_enter_exit_debug
async def async_migrate_entry(hass: HomeAssistant, config_entry: ElasticIntegrationConfigEntry) -> bool:
    """Handle migration of config entry."""
    if config_entry.version == ElasticFlowHandler.VERSION:
        return True

    try:
        migrated_data, migrated_options, migrated_version = migrate_data_and_options_to_version(
            config_entry,
            ElasticFlowHandler.VERSION,
        )
    except ValueError:
        LOGGER.exception(
            "Migration failed attempting to migrate from version %s to version %s. Ended on %s.",
            config_entry.version,
            ElasticFlowHandler.VERSION,
        )
        return False

    hass.config_entries.async_update_entry(
        config_entry,
        data=migrated_data,
        options=migrated_options,
        version=migrated_version,
    )

    return True


@log_enter_exit_debug
def migrate_data_and_options_to_version(
    config_entry: ElasticIntegrationConfigEntry,
    desired_version: int,
) -> tuple[dict, dict, int]:
    """Migrate a config entry from its current version to a desired version."""
    LOGGER.debug(
        "Migrating config entry from version %s to %s",
        config_entry.version,
        desired_version,
    )

    data = {**config_entry.data}
    options = {**config_entry.options}
    begin_version = config_entry.version
    current_version = begin_version

    if current_version < desired_version:
        for version in range(current_version + 1, desired_version + 1):
            migration_func = globals().get(f"migrate_to_version_{version}")
            if migration_func:
                data, options = migration_func(data, options)
                current_version = version

    end_version = current_version

    LOGGER.info("Migration from version %s to version %s successful", begin_version, end_version)

    return data, options, end_version


def migrate_to_version_2(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 2."""
    only_publish_changed = data.get("only_publish_changed", False)
    data["publish_mode"] = "All" if not only_publish_changed else "Any changes"

    if "only_publish_changed" in data:
        del data["only_publish_changed"]

    return data, options


def migrate_to_version_3(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 3."""
    if "health_sensor_enabled" in data:
        del data["health_sensor_enabled"]

    return data, options


def migrate_to_version_4(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 4."""
    if "index_mode" not in data:
        data["index_mode"] = "index"

    conf_ilm_max_size = "ilm_max_size"
    if conf_ilm_max_size in data:
        del data[conf_ilm_max_size]

    conf_ilm_delete_after = "ilm_delete_after"
    if conf_ilm_delete_after in data:
        del data[conf_ilm_delete_after]

    return data, options


def migrate_to_version_5(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 5."""
    keys_to_remove = [
        "datastream_type",
        "datastream_name_prefix",
        "datastream_namespace",
    ]

    for key in keys_to_remove:
        if key in data:
            del data[key]

    keys_to_migrate = [
        "publish_enabled",
        "publish_frequency",
        "publish_mode",
        "excluded_domains",
        "excluded_entities",
        "included_domains",
        "included_entities",
    ]

    for key in keys_to_migrate:
        if key not in options and key in data:
            options[key] = data[key]
        if key in data:
            del data[key]

    remove_keys_if_empty = [
        "username",
        "password",
        "api_key",
    ]

    for key in remove_keys_if_empty:
        if key in data and data[key] == "":
            del data[key]

    return data, options


def migrate_to_version_6(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 6."""

    if data.get("index_mode") is not None:
        del data["index_mode"]

    # Change publish mode to change_detection_type
    if options.get("publish_mode") is not None:
        if options["publish_mode"] == "All":
            options["polling_frequency"] = options["publish_frequency"]
            options["change_detection_type"] = ["STATE", "ATTRIBUTE"]

        if options["publish_mode"] == "Any changes":
            options["polling_frequency"] = 0
            options["change_detection_type"] = ["STATE", "ATTRIBUTE"]

        if options["publish_mode"] == "State changes":
            options["polling_frequency"] = 0
            options["change_detection_type"] = ["STATE"]

        del options["publish_mode"]

    else:
        options["polling_frequency"] = 0
        options["change_detection_type"] = ["STATE", "ATTRIBUTE"]

    # add dedicated settings for polling
    options_to_remove = [
        "ilm_enabled",
        "ilm_policy_name",
        "publish_mode",
        "publish_enabled",
        "index_format",
        "index_mode",
        "alias",
    ]

    for key in options_to_remove:
        if key in options:
            del options[key]

    return data, options


def migrate_to_version_7(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 7."""

    # if tags does not exist, set it to an empty array
    if "tags" not in options:
        options["tags"] = []

    options["targets_to_include"] = {}
    options["targets_to_exclude"] = {}
    options["exclude_targets"] = False
    options["include_targets"] = False

    if "included_entities" in options and len(options["included_entities"]) != 0:
        options["include_targets"] = True
        options["targets_to_include"]["entity_id"] = options["included_entities"]

    if "excluded_entities" in options and len(options["excluded_entities"]) != 0:
        options["exclude_targets"] = True
        options["targets_to_exclude"]["entity_id"] = options["excluded_entities"]

    keys_to_remove = [
        "excluded_domains",
        "excluded_entities",
        "included_domains",
        "included_entities",
    ]

    # Lowercase values in array options["change_detection_type"]
    if "change_detection_type" in options:
        options["change_detection_type"] = [x.lower() for x in options["change_detection_type"]]

    for key in keys_to_remove:
        if key in options:
            del options[key]

    return data, options
