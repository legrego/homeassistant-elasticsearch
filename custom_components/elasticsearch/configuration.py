"""Class for the Elastic integration config configuration objects."""

from dataclasses import dataclass

import voluptuous as vol
from elasticsearch.const import ONE_MINUTE, StateChangeType
from voluptuous import All, Exclusive, Inclusive, Optional, Required, Schema

DEFAULT_URL = "http://localhost:9200"

DEFAULT_PUBLISH_FREQUENCY = ONE_MINUTE
DEFAULT_POLLING_FREQUENCY = ONE_MINUTE

DEFAULT_CHANGE_DETECTION_ENABLED = True
DEFAULT_CHANGE_DETECTION_TYPE = [StateChangeType.STATE.value, StateChangeType.ATTRIBUTE.value]
DEFAULT_VERIFY_SSL = True
DEFAULT_TIMEOUT_SECONDS = 30


class Authentication:
    """Authentication for the Elastic integration."""


@dataclass
class ApiAuth(Authentication):
    """API Key authentication for the Elastic integration."""

    api_key: str | None


@dataclass
class BasicAuth(Authentication):
    """Basic authentication for the Elastic integration."""

    username: str | None
    password: str | None


@dataclass
class NoAuth(Authentication):
    """No authentication for the Elastic integration."""


class ElasticIntegrationConfig:
    """Data and options for the Elastic integration."""

    @dataclass
    class Data:
        """Data for the Elastic integration."""

        url: str
        timeout: int
        verify_ssl: bool
        ssl_ca_path: str | None
        authentication: Authentication

        def get_url_schema(self) -> dict[Required, type[str]]:
            """Get the URL schema."""
            return {vol.Required(self.url, default="http://localhost:9200"): str}

        def _get_validation_schema(self) -> Schema:
            """Validate the config entry."""

            return Schema(
                required=True,
                schema=All(
                    {
                        Required("url", default="test"): str,
                        Required("timeout"): int,
                        Required("verify_ssl"): bool,
                        Optional("ssl_ca_path"): str,
                        Optional("api_key"): str,
                        Optional("username"): str,
                        Optional("password"): str,
                    },
                    {
                        Inclusive("username", "password", msg="Both Username and Password must be set"),
                        Exclusive(
                            "api_key",
                            "username",
                            msg="Found API Key and Username, do not mix api key and username/password authentication",
                        ),
                        Exclusive(
                            "api_key",
                            "password",
                            msg="Found API Key and Password, do not mix api key and username/password authentication",
                        ),
                    },
                ),
            )

        def _auth_settings_to_object(self, auth_settings: dict) -> Authentication:
            """Create authentication object from settings."""
            if auth_settings.get("api_key") is not None:
                return ApiAuth(api_key=auth_settings["api_key"])

            if auth_settings.get("username") is not None and auth_settings.get("password") is not None:
                return BasicAuth(username=auth_settings["username"], password=auth_settings["password"])

            return NoAuth()

        def from_config_entry(self, config_entry) -> "ElasticIntegrationConfig.Data":
            """Create data from config entry."""
            data = config_entry.data
            validation_schema = self._get_validation_schema()
            validated_data = validation_schema(data)
            validated_data["authentication"] = self._auth_settings_to_object(validated_data)
            return ElasticIntegrationConfig.Data(**validated_data)

        def get_authentication_type(self) -> str:
            """Get the authentication type."""
            return self.authentication.__class__.__name__

        def get(self, key: str, default: str | None = None) -> str | None:
            """Get value from data."""
            return getattr(self, key, default)

    @dataclass
    class Options:
        """Options for the Elastic integration."""

        change_detection_enabled: bool
        change_detection_type: list[str]
        polling_frequency: int
        publish_frequency: int
        excluded_domains: list[str]
        excluded_entities: list[str]
        included_domains: list[str]
        included_entities: list[str]

        def get(self, key: str, default: str | None = None) -> str | None:
            """Get value from data."""
            return getattr(self, key, default)
