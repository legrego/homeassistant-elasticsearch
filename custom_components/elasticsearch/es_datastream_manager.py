"""Index management facilities."""

import json
from logging import Logger
from pathlib import Path

from custom_components.elasticsearch.es_gateway import ElasticsearchGateway

from .const import (
    DATASTREAM_METRICS_INDEX_TEMPLATE_NAME,
)
from .logger import LOGGER as BASE_LOGGER
from .logger import async_log_enter_exit_debug


class DatastreamManager:
    """Index management facilities."""

    _logger = BASE_LOGGER

    def __init__(
        self,
        gateway: ElasticsearchGateway,
        log: Logger = BASE_LOGGER,
    ) -> None:
        """Initialize index management."""

        self._logger = log

        self._gateway: ElasticsearchGateway = gateway

    @async_log_enter_exit_debug
    async def async_init(self) -> None:
        """Perform init for index management."""
        if await self._needs_index_template() or await self._needs_index_template_update():
            await self._create_index_template()

    @async_log_enter_exit_debug
    async def _needs_index_template(self) -> bool:
        matching_templates = await self._gateway.get_index_template(
            name=DATASTREAM_METRICS_INDEX_TEMPLATE_NAME,
            ignore=[404],
        )

        return len(matching_templates.get("index_templates", [])) == 0

    @async_log_enter_exit_debug
    async def _needs_index_template_update(self) -> bool:
        matching_templates = await self._gateway.get_index_template(
            name=DATASTREAM_METRICS_INDEX_TEMPLATE_NAME,
            ignore=[404],
        )

        matching_template = matching_templates.get("index_templates", [{}])[0]

        new_template = await self._get_index_template_from_disk()

        imported_version = matching_template["index_template"].get("version", 0)
        new_version = new_template.get("version", 0)

        if imported_version != new_version:
            self._logger.info(
                "Update required from [%s} to [%s] for Home Assistant datastream index template",
                imported_version,
                new_version,
            )
            return True

        return False

    @async_log_enter_exit_debug
    async def _create_index_template(self) -> None:
        """Initialize any required datastream templates."""

        with (Path(__file__).parent / "datastreams" / "index_template.json").open(
            encoding="utf-8",
        ) as json_file:
            index_template = json.load(json_file)

        action = "Creating" if (await self._needs_index_template()) else "Updating"
        self._logger.info("%s index template for Home Assistant datastreams", action)

        await self._gateway.put_index_template(
            name=DATASTREAM_METRICS_INDEX_TEMPLATE_NAME,
            body=index_template,
        )

        if action == "Updating":
            await self._rollover_ha_datastreams()

    @async_log_enter_exit_debug
    async def _rollover_ha_datastreams(self):
        """Rollover Home Assistant datastreams."""
        datastreams = await self._gateway.get_datastreams(datastream="metrics-homeassistant.*")

        for datastream in datastreams.get("data_streams", []):
            self._logger.info("Rolling over datastream [%s]", datastream["name"])
            await self._gateway.rollover_datastream(datastream=datastream["name"])

    @async_log_enter_exit_debug
    async def _get_index_template_from_disk(self) -> dict:
        """Retrieve the index template from disk."""
        with (Path(__file__).parent / "datastreams" / "index_template.json").open(
            encoding="utf-8",
        ) as json_file:
            return json.load(json_file)
