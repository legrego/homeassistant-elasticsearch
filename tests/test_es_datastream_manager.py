"""Tests for the index manager class."""
# noqa: F401 # pylint: disable=redefined-outer-name

from unittest.mock import AsyncMock, patch

import pytest
from custom_components.elasticsearch.es_datastream_manager import DatastreamManager


@pytest.fixture
async def datastream_manager(initialized_gateway):
    """Return an DatastreamManager instance."""
    return DatastreamManager(initialized_gateway)


class Test_DatastreamManager_Sync:
    """Test the DatastreamManager class sync methods."""

    async def test_init(self, datastream_manager, hass, gateway):
        """Test the __init__ method."""
        assert datastream_manager._gateway == gateway


@pytest.mark.asyncio
class Test_DatastreamManager_Async:
    """Test the DatastreamManager class async methods."""

    async def test_async_init(self, datastream_manager):
        """Test the async_init method."""
        # Mock the _create_index_template method
        with (
            patch.object(datastream_manager, "_create_index_template", AsyncMock()),
            patch.object(datastream_manager, "_needs_index_template", return_value=False),
            patch.object(datastream_manager, "_needs_index_template_update", return_value=False),
        ):
            await datastream_manager.async_init()

    async def test_needs_index_template(self, datastream_manager):
        """Test the _needs_index_template method."""

        # Mock the get_index_template method to return a matching template
        datastream_manager._gateway.get_index_template = AsyncMock(
            return_value={"index_templates": []},
        )

        result = await datastream_manager._needs_index_template()

        assert result

    @pytest.mark.asyncio
    async def test_create_index_template(self, datastream_manager, snapshot):
        """Test the _create_index_template method."""
        # Mock the _needs_index_template method to return False
        with patch.object(datastream_manager, "_needs_index_template", return_value=True):
            # Mock the put_index_template method
            datastream_manager._gateway.put_index_template = AsyncMock()

            await datastream_manager._create_index_template()

            datastream_manager._gateway.put_index_template.assert_called_once()

            call_args = datastream_manager._gateway.put_index_template.call_args.kwargs

            assert call_args == snapshot

    async def test_datastream_rollover(self, datastream_manager):
        """Test the datastream_rollover method."""
        # Mock the rollover_index method

        # GET /_data_stream/my-data-stream

        datastream_manager._gateway.rollover_datastream = AsyncMock()

        datastream_manager._gateway.get_datastreams = AsyncMock(
            return_value={
                "data_streams": [
                    {
                        "name": "metrics-homeassistant.sensor-default",
                        "indices": ["metrics-homeassistant.sensor-default"],
                        "data_stream": {
                            "name": "metrics-homeassistant.sensor-default",
                            "timestamp_field": {"name": "@timestamp"},
                        },
                    },
                    {
                        "name": "metrics-homeassistant.counter-default",
                        "indices": ["metrics-homeassistant.counter-default"],
                        "data_stream": {
                            "name": "metrics-homeassistant.counter-default",
                            "timestamp_field": {"name": "@timestamp"},
                        },
                    },
                ]
            }
        )

        await datastream_manager._rollover_ha_datastreams()

        datastream_manager._gateway.get_datastreams.assert_called_once()

        assert datastream_manager._gateway.rollover_datastream.call_count == 2

        datastream_manager._gateway.rollover_datastream.assert_called_with(
            datastream="metrics-homeassistant.counter-default"
        )

    async def test_needs_index_template_update(self, datastream_manager, snapshot):
        """Test the _needs_index_template_update method."""
        # Mock the get_index_template method to return a matching template
        datastream_manager._gateway.get_index_template = AsyncMock(
            return_value={
                "index_templates": [
                    {
                        "name": "datastream_metrics",
                        "index_template": {"version": 1},
                    }
                ]
            },
        )

        result = await datastream_manager._needs_index_template_update()

        assert result

    async def test_needs_index_template_no_update(self, datastream_manager, snapshot):
        """Test the _needs_index_template_update method."""
        # Mock the get_index_template method to return a matching template
        datastream_manager._gateway.get_index_template = AsyncMock(
            return_value={
                "index_templates": [
                    {
                        "name": "datastream_metrics",
                        "index_template": {"version": 2},
                    }
                ]
            },
        )

        result = await datastream_manager._needs_index_template_update()

        assert not result

    # @async_log_enter_exit_debug
    # async def _needs_index_template_update(self) -> bool:
    #     matching_templates = await self._gateway.get_index_template(
    #         name=DATASTREAM_METRICS_INDEX_TEMPLATE_NAME,
    #         ignore=[404],
    #     )

    #     matching_template = matching_templates.get("index_templates", [{}])[0]

    #     new_template = await self._get_index_template_from_disk()

    #     imported_version = matching_template["index_template"].get("version", 0)
    #     new_version = new_template.get("version", 0)

    #     if imported_version != new_version:
    #         self._logger.info(
    #             "Update required from [%s} to [%s] for Home Assistant datastream index template",
    #             imported_version,
    #             new_version,
    #         )
    #         return True

    #     return False
