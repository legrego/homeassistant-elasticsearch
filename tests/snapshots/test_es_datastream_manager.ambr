# serializer version: 1
# name: Test_DatastreamManager_Async.test_create_index_template[es8]
  dict({
    'body': dict({
      'composed_of': 'metrics-homeassistant@custom',
      'data_stream': dict({
      }),
      'ignore_missing_component_templates': 'metrics-homeassistant@custom',
      'index_patterns': list([
        'metrics-homeassistant.*-default',
      ]),
      'priority': 500,
      'template': dict({
        'lifecycle': dict({
          'data_retention': '365d',
        }),
        'mappings': dict({
          'dynamic': 'false',
          'dynamic_templates': list([
            dict({
              'hass_entity_attributes': dict({
                'mapping': dict({
                  'fields': dict({
                    'float': dict({
                      'ignore_malformed': True,
                      'type': 'float',
                    }),
                    'keyword': dict({
                      'ignore_above': 1024,
                      'type': 'keyword',
                    }),
                  }),
                  'type': 'text',
                }),
                'path_match': 'hass.entity.attributes.*',
              }),
            }),
          ]),
          'properties': dict({
            '@timestamp': dict({
              'format': 'strict_date_optional_time_nanos',
              'type': 'date_nanos',
            }),
            'agent': dict({
              'properties': dict({
                'ephemeral_id': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
                'id': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
                'name': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
                'type': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
                'version': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
              }),
            }),
            'data_stream': dict({
              'properties': dict({
                'dataset': dict({
                  'type': 'constant_keyword',
                }),
                'namespace': dict({
                  'type': 'constant_keyword',
                }),
                'type': dict({
                  'type': 'constant_keyword',
                  'value': 'metrics',
                }),
              }),
            }),
            'ecs': dict({
              'properties': dict({
                'version': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
              }),
            }),
            'event': dict({
              'properties': dict({
                'action': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
                'kind': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
                'type': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
              }),
            }),
            'hass': dict({
              'properties': dict({
                'entity': dict({
                  'properties': dict({
                    'area': dict({
                      'properties': dict({
                        'floor': dict({
                          'properties': dict({
                            'id': dict({
                              'type': 'keyword',
                            }),
                            'name': dict({
                              'type': 'keyword',
                            }),
                          }),
                          'type': 'object',
                        }),
                        'id': dict({
                          'type': 'keyword',
                        }),
                        'name': dict({
                          'type': 'keyword',
                        }),
                      }),
                      'type': 'object',
                    }),
                    'attributes': dict({
                      'dynamic': True,
                      'type': 'object',
                    }),
                    'device': dict({
                      'properties': dict({
                        'area': dict({
                          'properties': dict({
                            'floor': dict({
                              'properties': dict({
                                'id': dict({
                                  'type': 'keyword',
                                }),
                                'name': dict({
                                  'type': 'keyword',
                                }),
                              }),
                              'type': 'object',
                            }),
                            'id': dict({
                              'type': 'keyword',
                            }),
                            'name': dict({
                              'type': 'keyword',
                            }),
                          }),
                          'type': 'object',
                        }),
                        'class': dict({
                          'type': 'keyword',
                        }),
                        'id': dict({
                          'type': 'keyword',
                        }),
                        'labels': dict({
                          'type': 'keyword',
                        }),
                        'name': dict({
                          'type': 'keyword',
                        }),
                      }),
                      'type': 'object',
                    }),
                    'domain': dict({
                      'type': 'keyword',
                    }),
                    'friendly_name': dict({
                      'type': 'keyword',
                    }),
                    'geo': dict({
                      'properties': dict({
                        'location': dict({
                          'type': 'geo_point',
                        }),
                      }),
                      'type': 'object',
                    }),
                    'id': dict({
                      'type': 'keyword',
                    }),
                    'labels': dict({
                      'type': 'keyword',
                    }),
                    'name': dict({
                      'type': 'keyword',
                    }),
                    'object': dict({
                      'properties': dict({
                        'id': dict({
                          'time_series_dimension': True,
                          'type': 'keyword',
                        }),
                      }),
                      'type': 'object',
                    }),
                    'platform': dict({
                      'type': 'keyword',
                    }),
                    'state': dict({
                      'properties': dict({
                        'class': dict({
                          'type': 'keyword',
                        }),
                      }),
                    }),
                    'unit_of_measure': dict({
                      'type': 'keyword',
                    }),
                    'value': dict({
                      'fields': dict({
                        'keyword': dict({
                          'ignore_above': 1024,
                          'type': 'keyword',
                        }),
                      }),
                      'type': 'text',
                    }),
                    'valueas': dict({
                      'properties': dict({
                        'boolean': dict({
                          'type': 'boolean',
                        }),
                        'date': dict({
                          'format': 'strict_date',
                          'type': 'date',
                        }),
                        'datetime': dict({
                          'type': 'date',
                        }),
                        'float': dict({
                          'ignore_malformed': True,
                          'type': 'float',
                        }),
                        'integer': dict({
                          'ignore_malformed': True,
                          'type': 'integer',
                        }),
                        'string': dict({
                          'fields': dict({
                            'keyword': dict({
                              'ignore_above': 1024,
                              'type': 'keyword',
                            }),
                          }),
                          'type': 'text',
                        }),
                        'time': dict({
                          'format': 'HH:mm:ss.SSSSSS||time||strict_hour_minute_second||time_no_millis',
                          'type': 'date',
                        }),
                      }),
                    }),
                  }),
                  'type': 'object',
                }),
              }),
              'type': 'object',
            }),
            'host': dict({
              'properties': dict({
                'architecture': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
                'geo': dict({
                  'properties': dict({
                    'location': dict({
                      'type': 'geo_point',
                    }),
                  }),
                }),
                'hostname': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
                'name': dict({
                  'ignore_above': 1024,
                  'type': 'keyword',
                }),
                'os': dict({
                  'properties': dict({
                    'name': dict({
                      'ignore_above': 1024,
                      'type': 'keyword',
                    }),
                  }),
                }),
              }),
            }),
            'tags': dict({
              'ignore_above': 1024,
              'type': 'keyword',
            }),
          }),
        }),
        'settings': dict({
          'codec': 'best_compression',
          'index.mode': 'time_series',
          'mapping': dict({
            'total_fields': dict({
              'limit': '10000',
            }),
          }),
        }),
      }),
      'version': 2,
    }),
    'name': 'metrics-homeassistant',
  })
# ---
